pkg/instancetype/preference/apply/cpu.go

```diff
@@ -0,0 +1,68 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+package apply
+
+import (
+	virtv1 "kubevirt.io/api/core/v1"
+	v1beta1 "kubevirt.io/api/instancetype/v1beta1"
+)
+
+func applyCPUPreferences(preferenceSpec *v1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
+	if preferenceSpec.CPU == nil || len(preferenceSpec.CPU.PreferredCPUFeatures) == 0 {
+		return
+	}
+	// Only apply any preferred CPU features when the same feature has not been provided by a user already
+	cpuFeatureNames := make(map[string]struct{})
+	for _, cpuFeature := range vmiSpec.Domain.CPU.Features {
+		cpuFeatureNames[cpuFeature.Name] = struct{}{}
+	}
+	for _, preferredCPUFeature := range preferenceSpec.CPU.PreferredCPUFeatures {
+		if _, foundCPUFeature := cpuFeatureNames[preferredCPUFeature.Name]; !foundCPUFeature {
+			vmiSpec.Domain.CPU.Features = append(vmiSpec.Domain.CPU.Features, preferredCPUFeature)
+		}
+	}
+}
+
+func GetPreferredTopology(preferenceSpec *v1beta1.VirtualMachinePreferenceSpec) v1beta1.PreferredCPUTopology {
+	// Default to PreferSockets when a PreferredCPUTopology isn't provided
+	preferredTopology := v1beta1.Sockets
+	if preferenceSpec != nil && preferenceSpec.CPU != nil && preferenceSpec.CPU.PreferredCPUTopology != nil {
+		preferredTopology = *preferenceSpec.CPU.PreferredCPUTopology
+	}
+	return preferredTopology
+}
+
+const defaultSpreadRatio uint32 = 2
```

Comment:
Move to top of file or into GetSpreadOptions?

Reply:
I can't find a reference but I was sure best practice in golang suggests to define constants close to their actual use and not just the top of a file. I do also want to use this in tests in the future so I'd rather keep this separate from `GetSpreadOptions` for the time being anyway.

Reply:
I'd rather keep it out of the scope of the func so it's at least usable by other callers and tests in the future.